[SelectionDAG] Introduce ISD::PTRADD

ritter-x2a opened the pull request:
This opcode represents the addition of a pointer value (first operand) and an integer offset (second operand). PTRADD nodes are only generated if the TargetMachine opts in by overriding TargetMachine::shouldPreservePtrArith().

The PTRADD node and respective visitPTRADD() function were adapted by @rgwott from the CHERI/Morello LLVM tree.
Original authors: @davidchisnall, @jrtc27, @arichardson.

The changes in this PR were extracted from PR #105669.

davidchisnall committed:
[SelectionDAG] Introduce ISD::PTRADD

This opcode represents the addition of a pointer value (first operand) and an
integer offset (second operand). PTRADD nodes are only generated if the
TargetMachine opts in by overriding TargetMachine::shouldPreservePtrArith().

The PTRADD node and respective visitPTRADD() function were adapted by @rgwott
from the CHERI/Morello LLVM tree.
Original authors: @davidchisnall, @jrtc27, @arichardson.

The changes in this PR were extracted from PR #105669.

Co-authored-by: Jessica Clarke <jrtc27@jrtc27.com>
Co-authored-by: Alexander Richardson <alexrichardson@google.com>
Co-authored-by: Rodolfo Wottrich <rodolfo.wottrich@arm.com>

ritter-x2a committed:
Use AddToWorklist() instead of manually re-visiting a newly generated
node in the DAGCombines.
Also: remove else after return and braces around single-line blocks to
match the coding standards.

ritter-x2a commented:
The changes in the first commit are taken verbatim from #105669, the second one contains two minor fixes by me.

davidchisnall commented:
This looks very much like code I wrote (sorry). Tagging @resistor to check that it still looks like code that does the right thing for us.

resistor reviewed (commented):
Overall LGTM and matches what we have in the cheriot downstream: https://github.com/CHERIoT-Platform/llvm-project/tree/cheriot-upstream

I have noted a couple of places where I'd like to understand some optimization divergences from what we have, but I don't think those are blockers.

That said, I would like to hear from at least one other relevant reviewer before this goes in.

ritter-x2a commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -2617,6 +2620,100 @@ SDValue DAGCombiner::foldSubToAvg(SDNode *N, const SDLoc &DL) {
+
+  if (N0.getOpcode() == ISD::PTRADD &&
+      !reassociationCanBreakAddressingModePattern(ISD::PTRADD, DL, N, N0, N1)) {
+    SDValue X = N0.getOperand(0);
```
@rgwott provided explanations for changes in the DAGCombiner in the original PR, from which this is extracted:
- https://github.com/llvm/llvm-project/pull/105669#discussion_r1727485653
- https://github.com/llvm/llvm-project/pull/105669#discussion_r1727494906

Or, if you're referring to the use of `AddToWorklist(Add.getNode())` to let the worklist algorithm handle the new node instead of handling it manually: That way seems to be more consistent with the other DAG combines, where `visit` is almost never called directly (see here for previous discussion: https://github.com/llvm/llvm-project/pull/105669/files#r2065675292 ).

resistor replied:
No issues with the wordlist change. It seems like @jrtc27 's comments re: the changes to the DAG combines still need to be addressed?

ritter-x2a replied:
Which comments do you mean? The `(ptradd (ptradd x, y), z) -> (ptradd (ptradd x, z), y)` combine, which the first PR comment I linked refers to, has already been replaced with the code comment [starting at line 2669](https://github.com/llvm/llvm-project/pull/140017/files#diff-d0eb75096db76ab253fc7f8ae6343c4b4516fc619d851898cbdac1a5bf481941R2669-R2696) as described in the discussion below that PR comment.

resistor replied:
I was looking at this link where it didn't appear to have a conclusion, but perhaps I missed a cross-thread connection: https://github.com/llvm/llvm-project/pull/105669#discussion_r1727494906

ritter-x2a replied:
I see; I took that one to be concluded with the explanation that it broke a test case that was shared, but I can do a bit more debugging to see if I find what's wrong there.

ritter-x2a replied:
Having taken a look at the code path in the [CHERIoT](https://github.com/CHERIoT-Platform/llvm-project/blob/ba63f5a2d112f961b921f19656da7ac61fa1fd1c/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L2684) repo, I agree that the version there looks problematic. It creates an `ADD` node (or retrieves an equal existing one that has previously been memoized and could be used elsewhere), calls `visit` outside of the worklist algorithm (which might RAUW it) and then potentially does not use the new node, and doesn't add its users to the worklist.
Unless I'm missing something, it doesn't look like "perform this combine if it enables another combine" is a pattern that the DAGCombiner is designed for.

Other than that, the current version misses the "y and z are constant" case present in the CHERIoT version. I was planning to implement that in the subsequent PRs to enable PTRADD for AMDGPU and keep this PR as faithful as possible to @rgwott's, but I can also add it here already if that is preferred.

jrtc27 replied:
Yeah the "y + z can be constant folded" is the one I'm concerned about losing. Part of the problem here is this is a one-way transformation (going backwards can mean taking a pointer out of bounds that wasn't previously, which is a problem for both CPA and, much more extensively, CHERI), and we don't want to hide a constant operand behind an ADD node because then that can't be selected to an immediate offset by the backend, instead it has to compute the full offset into a register first. The idea behind the CHERI code is that it can peek into what the result would look like if it did the transformation and commit to doing it only after having more information. Yes, "y and z are both constants" will catch some cases, but that won't catch more complex arithmetic expressions that can be simplified (as a trivial example, `(ptradd (ptradd x, c), (sub 0, c))`, but that could be rather more obfuscated and require more peeking, which visitADD already knows how to do). Maybe the only way to do that correctly though is repeat a bunch of visitADD(Like) for visitPTRADD.

arichardson replied:
Could we do this as a follow up and land the basic version first? I don't want to stall this PR just to get it perfect the first time. We could file an issue for the missing fold linking to this discussion?

arichardson commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -2617,6 +2620,85 @@ SDValue DAGCombiner::foldSubToAvg(SDNode *N, const SDLoc &DL) {
+/// Try to fold a pointer arithmetic node.
+/// This needs to be done separately from normal addition, because pointer
+/// addition is not commutative.
+/// This function was adapted from DAGCombiner::visitPTRADD() from the Morello
```
I'm not sure we need this comment here, the git blame commit message should have all that information. We don't so this for other changes that came from other downstream forks.

ritter-x2a replied:
[Removed.](https://github.com/llvm/llvm-project/pull/140017/commits/f6a4cda119e7825a571b24feaa21223a86103d15)

arichardson reviewed (approved):
LGTM, but I'd like @jrtc27 to confirm the fold part looks good now.

ritter-x2a committed:
Remove DAGCombine comment referring to Morello and CHERI

jrtc27 commented on llvm/include/llvm/Target/TargetMachine.h:
```diff
@@ -468,6 +468,11 @@ class TargetMachine {
+  /// True if target has some particular form of dealing with pointer arithmetic
+  /// semantics. False if pointer arithmetic should not be preserved for passes
+  /// such as instruction selection, and can fallback to regular arithmetic.
+  virtual bool shouldPreservePtrArith(const Function &F) const { return false; }
```
Can this please include the pointer type being operated on? For CHERI we don't want to add PTRADD for integers if the target doesn't otherwise use them, only for capabilities.

ritter-x2a replied:
Added with d5f083ed1dadf2cbfa914c8f886059120f8a7ee0 since I don't see harm in adding the pointer's VT if it's used somewhere. We don't currently need it for AMDGPU; my goal there is just to have DAG nodes that [can meaningfully carry an `inbounds` flag](https://github.com/llvm/llvm-project/pull/131862).

jrtc27 commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -2388,7 +2391,7 @@ static bool canFoldInAddressingMode(SDNode *N, SDNode *Use, SelectionDAG &DAG,
 
   TargetLowering::AddrMode AM;
-  if (N->getOpcode() == ISD::ADD) {
+  if (N->getOpcode() == ISD::ADD || N->getOpcode() == ISD::PTRADD) {
```
It's present in Morello LLVM. No reason not to include this IMO, it makes this change more universally applicable.

ritter-x2a committed:
Add the pointer's EVT to the signature of TM::shouldPreservePtrArith()

arsenm commented on llvm/include/llvm/Target/TargetMachine.h:
```diff
@@ -469,9 +470,13 @@ class TargetMachine {
+  /// arithmetic should not be preserved for passes such as instruction
+  /// selection, and can fallback to regular arithmetic.
+  virtual bool shouldPreservePtrArith(const Function &F,
+                                      const EVT &PtrVT) const {
```
EVT should be passed by value. But also I don't see how this can be an EVT and not the original IR type (or just PointerType)

ritter-x2a replied:
I didn't pass it by value to avoid including the `ValueTypes.h` header in this header (and the potential compile-time cost), but I can change that.

We don't have the IR type in `SelectionDAG::getMemBasePlusOffset`, where this function is called (unless I'm missing a way to retrieve it from an SDValue). That function is not only used to lower GEPs in the builder, but also for places in legalization (e.g., for loads and stores of too-wide vector types), where there isn't necessarily a matching IR type.

arsenm replied:
This belongs in TargetLowering, not TargetMachine

ritter-x2a replied:
Moved it to `TargetLoweringBase` and changed the signature to pass the EVT by value in 759f0656a9bc4e50bdc6add4d218b9425120d8cc.

arsenm commented on llvm/include/llvm/Target/TargetMachine.h:
```diff
@@ -469,9 +470,13 @@ class TargetMachine {
-  /// semantics. False if pointer arithmetic should not be preserved for passes
-  /// such as instruction selection, and can fallback to regular arithmetic.
-  virtual bool shouldPreservePtrArith(const Function &F) const { return false; }
+  /// semantics for pointers with the given value type. False if pointer
+  /// arithmetic should not be preserved for passes such as instruction
+  /// selection, and can fallback to regular arithmetic.
```
Fundamentally I don't see why we would want this. We should just always useptr add. Is the plan to remove this when more code handles ptradd? Should document that

ritter-x2a replied:
Unless we add a DAGCombine that replaces ptradd with add (which would need a similar option to disable for backends that want to handle ptradd properly), modifying existing backends to use ptradd without causing regressions is a significant effort (that's at least my experience from working on it for AMDGPU).
Should I add an "This should be removed when PTRADD nodes are widely supported by backends." comment to the function?

arsenm replied:
Yes

ritter-x2a replied:
Added the comment in 759f0656a9bc4e50bdc6add4d218b9425120d8cc.

ritter-x2a committed:
Move shouldPreservePtrArith from TargetMachine to TargetLoweringBase,

add a comment to state that it should be removed eventually, and reorder
if/else cases to avoid a negated condition where shouldPreservePtrArith
is called.

ritter-x2a committed:
Add (YIsConstant && ZIsConstant) case to the DAGCombine.

arsenm commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1502,6 +1502,11 @@ enum NodeType {
   // Outputs: [rv], output chain, glue
   PATCHPOINT,
 
+  // PTRADD represents pointer arithmatic semantics, for targets that opt in
```
```suggestion
  // PTRADD represents pointer arithmetic semantics, for targets that opt in
```

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5641,7 +5641,8 @@ bool SelectionDAG::isADDLike(SDValue Op, bool NoWrap) const {
 bool SelectionDAG::isBaseWithConstantOffset(SDValue Op) const {
   return Op.getNumOperands() == 2 && isa<ConstantSDNode>(Op.getOperand(1)) &&
-         (Op.getOpcode() == ISD::ADD || isADDLike(Op));
+         (Op.getOpcode() == ISD::ADD || Op.getOpcode() == ISD::PTRADD ||
```
Probably should add an isPtrAdd helper for this add || ptradd pattern

arsenm commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1502,6 +1502,11 @@ enum NodeType {
+  // PTRADD represents pointer arithmatic semantics, for targets that opt in
+  // using shouldPreservePtrArith().
+  // ptr = PTRADD ptr, offset
+  PTRADD,
```
Should also cover this in SelectionDAGCompat

ritter-x2a replied:
An entry for that [already exists in SelectionDAGCompat](https://github.com/llvm/llvm-project/blob/416b7dfaa0d114b552c596d320f0aaac5651e61e/llvm/include/llvm/Target/GlobalISel/SelectionDAGCompat.td#L63), to generate `V_ADD3_U32_e64` for AMDGPU in GISel via a SelectionDAG rule.

arsenm replied:
Oh no, it's defined like this:

`def ptradd     : SDNode<"ISD::ADD"       , SDTPtrAddOp, []>;`

Since this now can map to 2 values, we'll need custom handling in GlobalISelEmitter

ritter-x2a replied:
Currently, this PR just changes this to
```
def ptradd : SDNode<"ISD::PTRADD" , SDTPtrAddOp, []>;
```
When looking through the related code, I didn't have the impression that it mattered to the use in GlobalISel what the SDAG node here is, changing it to `ISD::PTRADD` at least doesn't break the test for the `V_ADD3_U32_e64` matching.

ritter-x2a committed:
Add ISD::isPtrAdd() and fix a typo.

arsenm commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1729,6 +1734,10 @@ inline bool isExtVecInRegOpcode(unsigned Opcode) {
          Opcode == ISD::SIGN_EXTEND_VECTOR_INREG;
 }
 
+inline bool isPtrAdd(unsigned Opcode) {
+  return (Opcode == ISD::ADD) || (Opcode == ISD::PTRADD);
+}
```
I mean put this on SDValue, like isUndef

arichardson replied:
Yes that would be good. I also wonder if the name `isAddition`/`isAnyAdd` would work better since every ISD::PTRADD is an addition, but with `isPtrAdd`, not every ISD::ADD is a ptradd?

ritter-x2a replied:
Moved it to SDNode/SDValue like `isUndef`, and renamed it to `isAnyAdd` (as that name captures the semantics best to me).

topperc commented:
Should we add these verification asserts to `getNode` like we have for `ISD::ADD`

```
    assert(VT.isInteger() && "This operator does not apply to FP types!");       
    assert(N1.getValueType() == N2.getValueType() &&                             
           N1.getValueType() == VT && "Binary operator types must match!");
```

arichardson commented:
> Should we add these verification asserts(or some equivalent) to `getNode` like we have for `ISD::ADD`
> 
> ```
>     assert(VT.isInteger() && "This operator does not apply to FP types!");       
>     assert(N1.getValueType() == N2.getValueType() &&                             
>            N1.getValueType() == VT && "Binary operator types must match!");
> ```

I agree having these assertions would be useful, and we can relax them once PTRADD starts using actual pointer types instead of integer ones.

ritter-x2a commented:
> > Should we add these verification asserts(or some equivalent) to `getNode` like we have for `ISD::ADD`
> > ```
> >     assert(VT.isInteger() && "This operator does not apply to FP types!");       
> >     assert(N1.getValueType() == N2.getValueType() &&                             
> >            N1.getValueType() == VT && "Binary operator types must match!");
> > ```
> 
> I agree having these assertions would be useful, and we can relax them once PTRADD starts using actual pointer types instead of integer ones.

This raises the interesting question if we want to restrict the operand types to be the same for now, as it's done for ISD::ADD, or if we just want to require two possibly distinct integer types.
I don't think PTRADDs with different operand types would currently be constructed, since they are only constructed in places where we would also generate ISD::ADD nodes; requiring them to be the same would certainly make some advanced optimizations easier.

If PTRADD nodes with integer operands with different bitwidths should implement the GEP semantics for index types that are wider/narrower than the pointer type (i.e., the bits of the base pointer outside of the width of the index type are not affected by overflows in the index addition), this would add quite a footgun for adding future transformations of PTRADD nodes.

ritter-x2a committed:
Move isPtrAdd to SDNode/SDValue; rename it to isAnyAdd

ritter-x2a committed:
Fix formatting

ritter-x2a committed:
Add PTRADD handling to SDAG::getNode()

That involves early folds and (technically overly restrictive)
assertions on the operand types.

ritter-x2a commented:
1239537577ae406457928109d8d0cabe78ad588f enables the ADD verification assertions for PTRADD (with a comment to document why it is so overly restrictive) and a few straightforward early folds in `SelectionDAG::getNode()`.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -2617,6 +2620,85 @@ SDValue DAGCombiner::foldSubToAvg(SDNode *N, const SDLoc &DL) {
   return SDValue();
 }
 
+/// Try to fold a pointer arithmetic node.
```
Maybe should split this into a PR with tests and a target use

arsenm replied:
Yes, I think that probably the only way to do do this in manageable pieces

ritter-x2a replied:
Removed `DAGCombiner::visitPtrAdd` from this PR, to be added with tests and a target use to a separate later PR. For now, the split changes live on the [`isd-ptradd-initial-dagcombine` branch in my fork](https://github.com/ritter-x2a/llvm-project/tree/isd-ptradd-initial-dagcombine).

ritter-x2a committed:
Split the DAGCombiner::visitPtrAdd() implementation off to put it into a separate PR.

s-barannikov commented:
This patch adds more calls to `getMemBasePlusOffset()`, but the method wasn't updated to generate `ISD::PTRADD`, is this intended?

topperc commented:
> This patch adds more calls to `getMemBasePlusOffset()`, but the method wasn't updated to generate `ISD::PTRADD`, is this intended?

Wasn't it updated here?

```
SDValue SelectionDAG::getMemBasePlusOffset(SDValue Ptr, SDValue Offset,
                                           const SDLoc &DL,
                                           const SDNodeFlags Flags) {
  assert(Offset.getValueType().isInteger());
  EVT BasePtrVT = Ptr.getValueType();
  if (TLI->shouldPreservePtrArith(this->getMachineFunction().getFunction(),
                                  BasePtrVT))
    return getNode(ISD::PTRADD, DL, BasePtrVT, Ptr, Offset, Flags);
  return getNode(ISD::ADD, DL, BasePtrVT, Ptr, Offset, Flags);
}
```

topperc commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -9138,8 +9138,8 @@ bool SelectionDAGBuilder::visitMemPCpyCall(const CallInst &I) {
   // Adjust return pointer to point just past the last dst byte.
-  SDValue DstPlusSize = DAG.getNode(ISD::ADD, sdl, Dst.getValueType(),
-                                    Dst, Size);
+  SDNodeFlags Flags;
```
Doesn't the Flags argument of getMemBasePlusOffset have a default value?

ritter-x2a replied:
Indeed, I've now removed the unnecessary explicit argument setting here, thanks! I think we could actually use NUW here since this operation computes the position one-past-the-end of a memcpy'd memory range (which should either be within or one-past-the-end of an allocated object, which may not wrap according to the IR LangRef), but let's not add more functionality changes to this PR.

s-barannikov commented:
> > This patch adds more calls to `getMemBasePlusOffset()`, but the method wasn't updated to generate `ISD::PTRADD`, is this intended?
> 
> Wasn't it updated here?
> 
> ```
> SDValue SelectionDAG::getMemBasePlusOffset(SDValue Ptr, SDValue Offset,
>                                            const SDLoc &DL,
>                                            const SDNodeFlags Flags) {
>   assert(Offset.getValueType().isInteger());
>   EVT BasePtrVT = Ptr.getValueType();
>   if (TLI->shouldPreservePtrArith(this->getMachineFunction().getFunction(),
>                                   BasePtrVT))
>     return getNode(ISD::PTRADD, DL, BasePtrVT, Ptr, Offset, Flags);
>   return getNode(ISD::ADD, DL, BasePtrVT, Ptr, Offset, Flags);
> }
> ```

I looked for it three times =\

ritter-x2a committed:
Remove redundant flag argument
